python/lsst/sims/catUtils/mixins/sncat.py

```diff
@@ -166,44 +183,60 @@ def get_snbrightness(self):
             self.column_by_name('raJ2000'),\
             self.column_by_name('decJ2000')
 
-        SNobject = SNObject()
+        raDeg = np.degrees(ra)
+        decDeg = np.degrees(dec)
+
+        ebv = self.column_by_name('EBV')
+        id_list = self.column_by_name('snid')
+
         bandname = self.obs_metadata.bandpass
         if isinstance(bandname, list):
             raise ValueError('bandname expected to be string, but is list\n')
         bandpass = self.lsstBandpassDict[bandname]
 
-        # Initialize return array
-        vals = np.zeros(shape=(self.numobjs, 4))
-
-        for i, _ in enumerate(vals):
-            SNobject.set(z=_z[i], c=c[i], x1=x1[i], t0=t0[i], x0=x0[i])
-            SNobject.setCoords(ra=ra[i], dec=dec[i])
-            SNobject.mwEBVfromMaps()
+        # Initialize return array so that it contains the values you would get
+        # if you passed through a t0=self.badvalues supernova
+        vals = np.array([[0.0]*len(t0), [np.inf]*len(t0),
+                        [np.nan]*len(t0), [np.inf]*len(t0)]).transpose()
+
+        for i in np.where(np.logical_and(np.isfinite(t0), np.abs(self.mjdobs-t0)<self.maxTimeSNVisible))[0]:
+
+            if id_list[i] in self._sn_object_cache:
+                SNobject = self._sn_object_cache[id_list[i]]
+            else:
+                SNobject = SNObject()
+                SNobject.set(z=_z[i], c=c[i], x1=x1[i], t0=t0[i], x0=x0[i])
+                SNobject.setCoords(ra=raDeg[i], dec=decDeg[i])
+                SNobject.set_MWebv(ebv[i])
+                self._sn_object_cache[id_list[i]] = SNobject
```

Comment:
Just recording this as something to try later:

Instead of Caching SNObject instances, we could either keep pickled objects or SNObject.SNstate in some format. Currently SNstate is a dictionary and maybe changing it to a named tuple will decrease the size of the variable. The point is that SNObject instances can be reconstructed from SNstate via the 
```
snstate = snobject.SNstate # snobject is an instance of SNObject
s = SNObject.fromSNstate(snstate)
# s should have the same properties as snobject
```
This was the kind of purpose, these methods were built for.